web3.js/src/connection.ts

```diff
@@ -3039,104 +3086,290 @@ export class Connection {
-          try {
-            const blockHeight = await this.getBlockHeight(commitment);
-            return blockHeight;
-          } catch (_e) {
-            return -1;
+        throw new TransactionExpiredBlockheightExceededError(signature);
+      }
+    } finally {
+      done = true;
+      abortConfirmation();
+    }
+    return result;
+  }
+
+  private async confirmTransactionUsingDurableNonceStrategy({
+    commitment,
+    strategy: {minContextSlot, nonceAccountPubkey, nonceValue, signature},
+  }: {
+    commitment?: Commitment;
+    strategy: DurableNonceTransactionConfirmationStrategy;
+  }) {
+    let done: boolean = false;
+    const expiryPromise = new Promise<{
+      __type: TransactionStatus.NONCE_INVALID;
+    }>(resolve => {
+      let currentNonceValue: string | undefined = nonceValue;
+      const getCurrentNonceValue = async () => {
+        try {
+          const nonceAccount = await this.getNonce(nonceAccountPubkey, {
+            commitment,
+            minContextSlot,
+          });
+          return nonceAccount?.nonce;
+        } catch (e) {
+          // If for whatever reason we can't reach/read the nonce
+          // account, just keep using the last-known value.
+          return currentNonceValue;
+        }
+      };
+      (async () => {
+        currentNonceValue = await getCurrentNonceValue();
+        if (done) return;
+        while (
+          true // eslint-disable-line no-constant-condition
+        ) {
+          if (nonceValue !== currentNonceValue) {
+            resolve({__type: TransactionStatus.NONCE_INVALID});
+            return;
           }
-        };
-        (async () => {
-          let currentBlockHeight = await checkBlockHeight();
+          await sleep(2000);
+          if (done) return;
+          currentNonceValue = await getCurrentNonceValue();
           if (done) return;
-          while (currentBlockHeight <= config.lastValidBlockHeight) {
-            await sleep(1000);
-            if (done) return;
-            currentBlockHeight = await checkBlockHeight();
-            if (done) return;
+        }
+      })();
+    });
+    const {abortConfirmation, confirmationPromise} =
+      this.getTransactionConfirmationPromise({commitment, signature});
+    let result: RpcResponseAndContext<SignatureResult>;
+    try {
+      const outcome = await Promise.race([confirmationPromise, expiryPromise]);
+      if (outcome.__type === TransactionStatus.PROCESSED) {
+        result = outcome.response;
+      } else {
+        // Double check that the transaction is indeed unconfirmed.
+        let signatureStatus:
+          | RpcResponseAndContext<SignatureStatus | null>
+          | null
+          | undefined;
+        while (
+          true // eslint-disable-line no-constant-condition
+        ) {
+          const status = await this.getSignatureStatus(signature);
+          if (status == null) {
+            break;
+          }
+          if (status.context.slot < minContextSlot) {
```

Comment:
I think we should be checking against the context slot last returned by the `getNonce` request so that we know that if the nonce was changed, it definitely wasn't for the transaction we are trying to confirm. By using the original `minContextSlot` we could get `null` as a response from a delayed RPC node that will soon process the transaction.